### ISSUE TITLE
Audio delays, choppiness, etc.

### ISSUE THREAD

**Issue by richard42** [mupen64plus/mupen64plus-user-issues#227]
Originally reported on Google Code with ID 228

```
Ubuntu 8.10 (Intrepid Ibex) amd64

mupen64plus v1.5-trunk-r1384
video plugins: Rice 1.5-r1384, gln64 0.4.1, Glide64 (05/27/09 10am? That's
when I updated the svn..). (no response from z64gl - black screen)
audio plugin: blight sdl input 0.0.10 (someone tell me there's ANYTHING
else out there....)

Regardless of the video plugin used, there is a delay in audio for all
games (between 0.5 and 2.0 seconds). The audio plays most smoothly with
Glide64 and stutters quite a lot with gln64. Rice hits somewhere between
the two.

I've tried playing the buffer settings, but there's really no thorough
guide to how these settings work or what they do exactly. It seems to me
there's a trade off between getting smooth audio and getting it synced to
the video. The default settings make most games unplayable for me.

Is there a better way to do this? Are there any other audio plugins
available anywhere? Particularly I'm looking for one that would manually
sync audio to video or vice versa. (ie, by slowing one down for the other
to catch up or speeding one up for a short time)


```

Reported by `david%server.jp@gtempaccount.com` on 2009-05-27 02:54:29

**Comment by richard42** [comment]
```
type-o,
I listed the input plugin where I should have typed the audio plugin!

audio plugin: Jttl's SDL Audio 1.5-r1384 (need something better here...)
input plugin: blight's sdl input 0.0.10 (this plugin is just fine!)
```

Reported by `david%server.jp@gtempaccount.com` on 2009-05-27 02:59:27

**Comment by richard42** [comment]
```
I don't think that this is an audio plugin problem.  Pulse audio has been known to
cause a lot of issues; if you're using Pulse you might want to get rid of it and try
with just alsa.  Otherwise you should open up a system monitor window and look at
your CPU usage when the sound is stuttering.  In most cases this problem is caused
by
not enough CPU/GPU power to run the game at a real-time rate.
```

Reported by `richard42g` on 2009-05-28 01:18:26

**Comment by richard42** [comment]
```
I've got the same problem, using i915 video module (direct rendering enabled and
perfectly working). Pulseaudio presence (and absence) doesn't change anything. The
sound stays very choppy and laggy.

Mupen64plus uses around 60% of the CPU according to "top".
```

Reported by `victor.lavaud` on 2009-12-07 01:00:35

**Comment by richard42** [comment]
```
A week or two ago I completely re-wrote the audio buffering and synchronization code
in the SDL-audio plugin.  It did have a number of problems.  The values for primary
and secondary buffer sizes/thresholds were in different units, which made them hard
to understand and broken for certain use cases (fast-forward and changing input
frequencies). The "secondary buffer high threshold" was basically unused, because the
algorithm essentially used the low threshold as its target.  And, the default
settings for the primary buffer size and secondary buffer low threshold left no
buffering overhead at all, resulting in frequent pausing and resuming the SDL audio
device, and probably the choppiness you observed.  This problem was not evident on
PCs which were very fast (ie < 30% CPU load while running).

All of these problems have been fixed in the new audio code, which now supports a
proper secondary buffer fullness target parameter, allowing the user to select a
tradeoff between a/v latency and likelihood of choppiness, which can be tuned to the
speed of the host PC.

This code is available for beta testing as described on the google code home page,
and will be released with Mupen64Plus 2.0.
```

Reported by `richard42g` on 2010-01-12 20:26:58
- **Status changed**: `Fixed`
